Cython/Compiler/Main.py

```diff
@@ -599,23 +622,45 @@ def add(self, source, result):
         self.num_errors += result.num_errors
 
 
-def compile_single(source, options, full_module_name = None):
+def get_fingerprint(cache, source, options):
+        from ..Build.Dependencies import create_dependency_tree
+        from ..Build.Cache import FingerprintFlags
+        context = Context.from_options(options)
+        dependencies = create_dependency_tree(context)
+        return cache.transitive_fingerprint(
+                source, dependencies.all_dependencies(source), options,
+                FingerprintFlags(
+                    'c++' if options.cplus else 'c',
+                    np_pythran=options.np_pythran
+                )
+        )
+
+
+def compile_single(source, options, full_module_name, cache=None, context=None, fingerprint=None):
     """
-    compile_single(source, options, full_module_name)
+    compile_single(source, options, full_module_name, cache, context, fingerprint)
 
     Compile the given Pyrex implementation file and return a CompilationResult.
     Always compiles a single file; does not perform timestamp checking or
     recursion.
     """
-    return run_pipeline(source, options, full_module_name)
 
+    if context is None:
+        context = Context.from_options(options)
 
-def compile_multiple(sources, options):
+    if cache:
+        fingerprint = fingerprint if fingerprint else get_fingerprint(cache, source, options)
```

Comment:
A little too many fingerprints for my taste.
```suggestion
        fingerprint = fingerprint or get_fingerprint(cache, source, options)
```